### ISSUE TITLE
Doc: provide an usage example of combining fpdf2 with python-barcode

### ISSUE THREAD

**Issue by Lucas-C** [PyFPDF/fpdf2#720]
We already have a page about [Barcodes](https://pyfpdf.github.io/fpdf2/Barcodes.html) in our documentation.
It was recently reported that we don't support code 128 barcodes: https://github.com/PyFPDF/fpdf2/discussions/719

[python-barcode](https://python-barcode.readthedocs.io/en/stable/) seems like a great library that can produces images for many barcode formats, using Pillow.

It would be interesting to provide a documentation section on how to generate barcodes with `python-barcode` and embed them in PDFs using `fpdf2`. Priority should be given to documenting how to insert them as SVG, as it will produce smaller PDF documents thant embedding barcodes as rater images.

---

By implementing this feature you, as a benevolent [FLOSS](https://en.wikipedia.org/wiki/Free/libre_and_open-source_software) developper, will provide access to the large community of fpdf2 users to a useful PDF functionality.
As a contributor you will get review feedbacks from maintainers & other contributors, and learn about the lifecycle & structure of a Python library on the way.
You will also be added into the [contributors list & map](https://github.com/PyFPDF/fpdf2#contributors-).

This issue can count as part of [hacktoberfest](https://hacktoberfest.com)

**Comment by shindodkar** [comment]
I would like to work on this project.✔

**Comment by Lucas-C** [comment]
> I would like to work on this project.✔

Hi and welcome @shindodkar!

The first step to work on thi would be to read the [Development guidelines](https://pyfpdf.github.io/fpdf2/Development.html).

Regarding this task, the goal will be to submit a Pull Request:
* https://opensource.guide/how-to-contribute/#how-to-submit-a-contribution
* https://opensource.com/article/19/7/create-pull-request-github

This PR will be an update of the Markdown file https://github.com/PyFPDF/fpdf2/blob/master/docs/Barcodes.md.
The idea is to add a new section that explains how to combine both libraries, `fpdf2` & `python-barcode`.

Of course, prior to this Pull Request, you will have to experiment yourself on how to do such thing 😄 
This documentation page would be a good starting point: https://python-barcode.readthedocs.io/en/stable/getting-started.html#usage
`SVGWriter` should be the prefered approach, as it will generate smaller PDF documents.

Also: [No such thing as a stupid question](https://en.wikipedia.org/wiki/No_such_thing_as_a_stupid_question): feel free to ask anything there if things are unclear!

**Comment by Lucas-C** [comment]
Hi @shindodkar

I see that you are requesting help from other contributors:
* https://github.com/PyFPDF/fpdf2/pull/726#issuecomment-1484959492
* https://github.com/PyFPDF/fpdf2/issues/723#issuecomment-1484939989

Could you please wait for those users to answer before posting any more comments on this repo?
People can be slow to respond on GitHub, and that is totally fine.
Please stop solicitating help from users in comments until you get an answer from those original requests 😅

Also, I would advise to:
1. carefuly read my comment just above, in order to get a grasp of how things works by yourself
2. if you are still a bit lost, please take the time to redact preceise questions, and post them in a new discussion: https://github.com/PyFPDF/fpdf2/discussions

**Comment by ssavi-ict** [comment]
Hi @Lucas-C . How are you doing? 
I have gone through this Issue. I read the [Barcode](https://pyfpdf.github.io/fpdf2/Barcodes.html) page and tried to understand what are actually trying to say about this issue. What I understand from that page is -

1. Generating a `Code 39` is already included in the `fpdf2`. Hence the first example [`Code 39`](https://pyfpdf.github.io/fpdf2/Barcodes.html#code-39) shows directly converting a text to Barcode in `Code 39` format.
2. The same has been done for the [`Interleaved 2 of 5`](https://pyfpdf.github.io/fpdf2/Barcodes.html#interleaved-2-of-5) method too. 
3. For the example of [`PDF-417`](https://pyfpdf.github.io/fpdf2/Barcodes.html#pdf-417) the example has been used  `pdf417` package including the implication of `fpdf2`.
4. The same has been done for the `QRCode` and `DataMatrix` examples. 

Please correct me if I am wrong till now.

What I have understood by reading this issue is `To provide some documentation and sample code similar to the 3rd and 4th as I have mentioned above`.  

Am I correct? 
If it is I would like to give it a try. However, I am not sure about I have clearly understood this line **Priority should be given to documenting how to insert them as SVG, as it will produce smaller PDF documents than embedding barcodes as rater images.**

Thank You.

**Comment by Lucas-C** [comment]
> Am I correct?

Yes, you are right 😊
The goal is indeed to provide some documentation and sample code similar on how to use the `python-barcode` lib with `fpdf2`

> However, I am not sure about I have clearly understood this line: **Priority should be given to documenting how to insert them as SVG, as it will produce smaller PDF documents than embedding barcodes as rater images.**

As can be seen on the page https://python-barcode.readthedocs.io/en/stable/getting-started.html#usage
there are 2 main approaches to generate images with `python-barcode`:
1. `from barcode.writer import SVGWriter`
2. `using `from barcode.writer import ImageWriter`

What I meant is simply that the `fpdf2` documentation added to close this issue should **focus on the 1st approach, using `SVGWriter`** 😊

**Comment by ssavi-ict** [comment]
@Lucas-C , Thank you for the reply. 

I was trying to create a snippet that may fix this issue. However, before that, I tried the following snippet -

```
from fpdf import FPDF

pdf = FPDF()
pdf.add_page()
pdf.image("vector.svg")
pdf.output("doc-with-svg.pdf")
```

from [this example](https://pyfpdf.github.io/fpdf2/SVG.html#basic-usage) that converts an SVG image (`vector.svg`) into a PDF (`doc-with-svg.pdf`). It is showing the following errors - 
```
C:\Python310\python.exe C:\Users\user\Documents\GitHubRepo\BarCode\main.py 
Traceback (most recent call last):
  File "C:\Users\user\Documents\GitHubRepo\BarCode\main.py", line 20, in <module>
    pdf.image("vector.svg")
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\fpdf.py", line 262, in wrapper
    return fn(self, *args, **kwargs)
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\fpdf.py", line 3746, in image
    return self._vector_image(img, x, y, w, h, link, title, alt_text)
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\fpdf.py", line 3875, in _vector_image
    svg = SVGObject(img.getvalue())
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\svg.py", line 646, in __init__
    self.convert_graphics(svg_tree)
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\svg.py", line 696, in convert_graphics
    self.build_group(root_tag, base_group)
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\svg.py", line 888, in build_group
    pdf_group.add_item(self.build_group(child))
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\svg.py", line 892, in build_group
    pdf_group.add_item(getattr(ShapeBuilder, shape_tags[child.tag])(child))
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\svg.py", line 339, in rect
    width = resolve_length(tag.attrib.get("width", 0))
  File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\fpdf\svg.py", line 115, in resolve_length
    raise ValueError(
ValueError: 100% uses unsupported relative length %
```

Is there anything I need to apply before running this snippet? Is this a Bug? 
Should I report a Bug? To create a PDF with `from barcode.writer import SVGWriter` I faced the same issue. 

However, I tried the following snippet - 
```
from fpdf import FPDF
from barcode import Code128
from barcode.writer import ImageWriter

# Create a new PDF document
pdf = FPDF()
pdf.add_page()

code128_img = Code128("1234567890", writer=ImageWriter())
code128_img.save('vector')
pdf.image('vector.png')
pdf.output('new.pdf')
```

It worked fine. And I have successfully created a PDF named `new.pdf` with `FPDF`. 
Waiting for your kind response.

Thanks.

**Comment by Lucas-C** [comment]
Oh sorry, you already mentioned this code in your previous comment, I missed it 🤦‍♂️
I'll have a look at it this evening! 😊

**Comment by Lucas-C** [comment]
I fixed the `ValueError` you reported in https://github.com/PyFPDF/fpdf2/pull/744

You can check that your last code snippet, with the call to `SVGWriter`, works using the latest version of `fpdf2` from this repository `master` branch:

    pip install git+https://github.com/PyFPDF/fpdf2.git@master

**Comment by ssavi-ict** [comment]
Thanks for the Quick fix @Lucas-C , Works fine. Except for the following warning -
`UserWarning: <svg> has no "viewBox", using its "width" & "height" as default "viewBox"
  warnings.warn(`
Which I think can be Ignored.

However, I have created the 128 Barcode PDF using `fpdf` and am ready to push the changes to the `Barcode.md`. Could you please share in which files should I reflect the changes along with the `Barcode.md` file? 

Thank You.

**Comment by Lucas-C** [comment]
> Could you please share in which files should I reflect the changes along with the `Barcode.md` file?

Please just add a mention of this additon in the doc to `CHANGELOG.md`!

**Comment by ssavi-ict** [comment]
@Lucas-C Thank You. I will do it by tonight.

**Comment by ssavi-ict** [comment]
@Lucas-C , with these small contributions, I would like to contribute features in the Codebase as well. What are the relevant knowledge and learnings I should gather in case of contributing to including new features and/or resolving bugs?

Thank You.

**Comment by Lucas-C** [comment]
> @Lucas-C , with these small contributions, I would like to contribute features/resolve bugs in the Codebase as well. What are the relevant knowledge and learnings I should gather in case of contributing to including new features and/or resolving bugs?

Most information is on https://pyfpdf.github.io/fpdf2/Development.html

I also gave some general advice there: https://github.com/PyFPDF/fpdf2/issues/496#issuecomment-1271260909

**Comment by ssavi-ict** [comment]
Thank You @Lucas-C .
I think the `docs\` directory should be a bit tidy. 
All the images, demo py files, Tutorials, etc should have some designated locations like `image\`, `tutorials\`, 'demo_codes\' etc.

**Comment by Lucas-C** [comment]
> Thank You @Lucas-C . I think the `docs\` directory should be a bit tidy. All the `images`, `demo py files`, `Tutorials`, etc should have some designated locations like `image\`, `tutorials\`, `demo_codes\` etc.

Why not, this seems like a good idea to me 😊
Would you like to submit a PR that improves things as you suggest?

**Comment by ssavi-ict** [comment]
@Lucas-C , I will be more than happy to do this. It may take some time. 
And I will be needing your help doing this. As the documentation stuff depends on those files. 
Also, I think it will be good if we open a new thread to track these improvements.

**Comment by Lucas-C** [comment]
Thanks to @ssavi-ict, a documentation section has now been added:
https://pyfpdf.github.io/fpdf2/Barcodes.html#code128

A bugfix was spotted in the process of redacting this documentation section,
and a fix has just been released in [v2.7.2](https://github.com/PyFPDF/fpdf2/releases/tag/2.7.2)
